clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/artifact/Replacer.java

```diff
@@ -136,7 +137,7 @@ private boolean replaceIfPossible(DocumentContext obj, Artifact artifact) {
   private static final Replacer DOCKER_IMAGE =
       builder()
           .path("$..spec.template.spec['containers', 'initContainers'].[?].image")
-          .replaceFilter(a -> filter(where("image").is(a.getName())))
+          .replaceFilter(a -> filter(where("image").regex(Pattern.compile(a.getName() + "(:.*)?"))))
```

Comment:
Is there a way for us to use `nameFromReference` to extract the name excluding the colon + tag sequence, or do we need to use a regex to match?

Reply:
Regex match was the only way I could see for us to find the image, with or without tag, in the manifest given the name of the image that is replacing.

Reply:
Okay! Just felt it was worth calling out as it feels a little strange at first glance to use a regex in `replaceFilter` when we have kind of similar logic in the `nameFromReference` method to extract just the artifact name from name + optional tag that doesn't use a regex, but I trust your judgment 👍 

One weird edge case that came to mind is an image that contains both a colon-delimited tag and a colon-delimited port (in which case we want to replace), or a colon-delimited digest and a colon-delimited port (in which case we do not want to replace); might be worth adding test coverage if we don't have already.

Reply:
Please feel free to correct me if I'm wrong though! From my understanding, ```nameFromReference``` is used to get the name of an image in a manifest, it's not useful here because we already know the image name that we want to replace from the artifact that is replacing (```a.getName()```). Instead we're looking to find an image in a manifest that has the same name as the artifact that is replacing, but with the caveat that the image in the manifest may have a tag on it.

With respect to the edge case, could you provide some more context or examples? I'm not familiar with the image formatting.

Reply:
+1 to Maggie's comment, doing this with a regex seems pretty fragile to me.  It also looks like this regex is just re-implementing `nameFromReference` though actually slightly differently (as `nameFromReference` handles images with digests not just tags).

At the very least we need much more robust tests (similar to [these](https://github.com/spinnaker/clouddriver/blob/faf0dfee31ae50b2e76808fa89cad068f1825448/clouddriver-kubernetes/src/test/java/com/netflix/spinnaker/clouddriver/kubernetes/artifact/ArtifactReplacerTest.java#L93)) as the single test case we have now is not sufficient to handle the added complexity of a regex here.

But ultimately I'm confident there's a way to lightly refactor this so that we can re-use `nameFromReference` instead of needing to re-implement it.  Obviously this isn't possible directly at the call site because we're just building an object.

But what if `replaceFilter` were actually be a `Function<String, JsonPath>` as it looks like we always only pass in the `artifact.getName()`.  Then we could call `nameFromReference` on `artifact.getName()` before passing it in.  I think that (or something similar) would work; while it's a tiny bit more code it would avoid having the existing logic re-implemented as a regex.

Reply:
I wrote my comment before seeing your comment...the tests I linked should have good examples of the image formatting.  What `nameFromReference` does is remove any digest/tag from an image, so we just have the actual image name.  It's used when reading artifacts from a manifest because we want the name to not have the tag (while the reference does).

But it could also be used when replacing because the manifest contains a full reference (though maybe it doesn't have a tag).  We effectively want to take that full reference in the manifest and figure out what the name is, then replace based on that name.  The regex is one solution (by ignoring anything after `:`) but `nameFromReference` is a well-tested and more robust way of doing that; if the manifest lists `gcr.io/my-project/my-image:tag` then `nameFromReference` will give us back `gcr.io/my-project/my-image` which is exactly what we want.

Reply:
> But what if replaceFilter were actually be a Function<String, JsonPath> as it looks like we always only pass in the artifact.getName(). Then we could call nameFromReference on artifact.getName() before passing it in

If my understanding is correct, we would want to  call ```nameFromReference``` on what's in the manifest image path not on ```artifact.getName()``` because ```artifact``` is the replacement and we already know the name, it's what ```artifact.getName()``` returns.